firedrake/mg/ufl_utils.py

```diff
@@ -193,7 +207,18 @@ def inject_on_restrict(fine, restriction, rscale, injection, coarse):
         V.dm.addCoarsenHook(None, inject_on_restrict)
 
     # Build set of coefficients we need to coarsen
-    forms = (problem.F, problem.J, problem.Jp)
+    forms = [problem.F, problem.J, problem.Jp]
+    # We need forms from EquationBCs, too.
+    # EquationBC can have an EquationBC on the boundary,
+    # so we need to keep looping until we've treated them all
+    bcs_to_consider = list(problem.bcs)
+    while len(bcs_to_consider) > 0:
+        bc = bcs_to_consider.pop()
+
+        if isinstance(bc, EquationBC):
+            forms += [bc._F.f, bc._J.f, bc._Jp.f]
+            bcs_to_consider += bc._F.sorted_equation_bcs()
+
```

Comment:
If we update `coefficient_mapping` dict in `coarsen_form()` and in `coarsen_function()`, I think we do not need to manually unfold forms inside `EquationBC` here (ll.210-227) as `coarsen_equation_bc` will recursively coarsen those forms using `coarsen_form`.